cluster-autoscaler/proposals/metrics.md

```diff
@@ -0,0 +1,90 @@
+## Cluster Autoscaler Monitoring
+
+## Introduction
+Currently the options to monitor Cluster Autoscaler are limited to logs, status
+configmap and events. While useful for debugging, none of this options is
+paritcularly practical for monitoring Cluster Autoscaler behaviour over time.
+This document describes a set of metrics (in Prometheus format) that will be
+added to 0.6 Cluster Autoscaler to allow better monitoring of its operations.
+
+## Current state
+Cluster Autoscaler already has a metrics endpoint providing some basic metrics.
+This includes default process metrics (number of goroutines, gc duration, cpu
+and memory details, etc) as well as some custom metrics related to time taken by
+various parts of Cluster Autoscaler main loop. Existing metrics will be renamed
+and refactored and new metrics will be added.
+
+## Metrics
+Cluster Autoscaler needs to monitor some aspects of cluster state as part of its normal operations. It can therefore provide metrics describing its own operations as well as some metrics related to general state of the cluster.
+
+### Cluster state
+* __cluster_healthy (gauge)__ - CA stops all operations if significant number of nodes
+are unready (by default 33% as of CA 0.5.4). This metric returns 1 if CA
+believes cluster to be healthy enough to continue normal operation, 0 otherwise.
+* __nodes_count (gauge)__ - number of nodes in different states in cluster. This
+metric is labeled with different node states:
+  * ready,
+  * unready,
+  * notStarted.
+* __unschedulable_pods_count (gauge)__ - number of unschedulable (“Pending”) pods in
+the cluster.
+
+### Cluster Autoscaler execution
+This metrics are refactored from currently existing metrics and track execution
+of various parts of Cluster Autoscaler loop.
+
+* __last_activity (gauge)__ - last time certain part of cluster autoscaler logic
+executed. Represented with unix timestamp. Uses the following set of labels:
+  * main - main loop iteration started.
+  * autoscaling - current state of the cluster has been updated, started autoscaling
+logic.
+  * scaleup - autoscaler will check if scale up is necessary.
+  * scaledown - autoscaler will try to scale down some nodes.
+  Not all of the above have to happen in every loop. For example if CA adds a new
+node it will completely skip scale down logic in this loop.
+* __duration_microseconds (summary)__ - time taken by various parts of CA loop. Uses
+the following set of labels:
+  * main - time of the whole iteration of main loop.
+  * updateClusterState - time used by CA to get node status from API server and
+update internal data structures.
+  * scaleup - time used to check if new node are necessary and add them.
+  * findUnneeded - time required to find nodes that are candidates for removal.
+  * scaledown - time required to verify unneeded nodes are really unnecessary and
```

Comment:
I'm assuming that labeling this metric like these allows us to e.g. selectively draw graphs of partial or total durations including one for total durations over time, one for main-loop durations over time. Am I correct?